datastore/datastore.go

```diff
@@ -0,0 +1,260 @@
+					fmt.Println(r.err)
+					continue
+				}
+
+				recordFindProvsComplete(ctx, r.status, metrics.FindProvsDuration.M(float64(r.duration)))
+
+				// if we failed to find this key, add to the failure cache so we don't try again for a bit
+				if r.status == "failed" {
+					failedLock.Lock()
+					if !failed.Has(r.key.String()) {
+						failed.Add(r.key.String())
+					}
+					failedLock.Unlock()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return func(q query.Query, res query.Results, err error) (query.Results, error) {
+		if err != nil {
+			return res, err
+		}
+
+		k := datastore.NewKey(q.Prefix)
+
+		// not interested if this is not a query for providers of a particular peer
+		// we're looking for /providers/cid, not /providers (currently used in GC)
+		if !providersRoot.IsAncestorOf(k) || len(k.Namespaces()) < 2 {
+			return res, err
+		}
+
+		failedLock.Lock()
+		if failed.Has(k.String()) {
+			failedLock.Unlock()
+			recordFindProvsComplete(ctx, "discarded")
+			return res, err
+		}
+		failedLock.Unlock()
+
+		var count int
+		onAfterNextSync := func(r query.Result, ok bool) (query.Result, bool) {
+			if ok {
+				count++
+				return r, ok
+			}
+			if count > 0 { // query has ended and there were locally found records
+				recordFindProvsComplete(ctx, "local")
+				return r, ok
+			}
+
+			pendingLock.Lock()
+			isPending, _ := pending[k.String()]
+			if !isPending {
+				// send to the find provs queue, if channel is full then discard...
+				select {
+				case findC <- k:
+					pending[k.String()] = true
+					stats.Record(ctx, metrics.FindProvsQueueSize.M(1))
+				case <-ctx.Done():
+				default:
+					recordFindProvsComplete(ctx, "discarded")
+				}
+			}
+			pendingLock.Unlock()
+
+			return r, ok
+		}
+
+		return hres.NewResults(res, hropts.OnAfterNextSync(onAfterNextSync)), nil
+	}
+}
+
+func findProviders(ctx context.Context, findC chan datastore.Key, foundC chan findResult, getRouting GetRoutingFunc, opts Options) {
+	done := func(r findResult) {
+		select {
+		case foundC <- r:
+		case <-ctx.Done():
+		}
+	}
+
+	for {
+		select {
+		case k := <-findC:
+			cid, err := providerKeyToCID(k)
+			if err != nil {
+				done(findResult{key: k, err: fmt.Errorf("failed to create CID from provider record key: %w", err)})
+				continue
+			}
+
+			routing, addProvider, err := getRouting(cid)
+			if err != nil {
+				done(findResult{key: k, err: fmt.Errorf("failed to get routing for CID: %w", err)})
+				continue
+			}
+
+			status := "failed"
+			start := time.Now()
+			fctx, cancel := context.WithTimeout(ctx, opts.FindProvidersTimeout)
+
+			for ai := range routing.FindProvidersAsync(fctx, cid, opts.FindProvidersCount) {
+				addProvider(ctx, cid, ai.ID)
+				status = "succeeded"
+			}
+
+			cancel()
+			done(findResult{key: k, status: status, duration: time.Since(start) / 1e+9})
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+var errInvalidKeyNamespaces = fmt.Errorf("not enough namespaces in provider record key")
```

Comment:
"enough" or just invalid?

Reply:
It is invalid because there are not enough - we expect at least 2 to be able to turn it into a CID

Reply:
Ah! Ok, I never thought of a subnamespace as "2 name spaces", hence my confusion. All good.

Reply:
Gotcha it’s just because the key is /providers/CID/PeerID - The call to Namespaces() returns [providers, CID, PeerID]